### ISSUE TITLE
negative/low FST values uncorrelated with other FST estimates, PCangsd

### ISSUE THREAD

**Issue by TonyKess** [ANGSD/angsd#346]
Hi again,

I know this issue has been raised in a few other questions, but I continue to run into the problem of low/negative FST estimates. Poolseq analysis of my populations, and independent comparison with a SNP chip using larger sample sizes show correlated FSTs across regions, and these also show very high correspondence with pcangsd PC loadings, both per locus and overall, and replicate results from a microsat study. But FST from the sfs step continues to give strange results  I am working with two relatively small samples (10 per pop). 
My input is as follows: 
` angsd/misc/realSFS POP1.saf.idx POP2.saf.idx -fold 1 > POP1.POP2.sfs
angsd/misc/realSFS fst index  POP1.saf.idx POP2.saf.idx -sfs POP1.POP2.sfs -fold 1 -fstout POP1.POP2
angsd/misc/realSFS fst stats POP1.POP2.fst.idx -fold 1
angsd/misc/realSFS fst stats2 POP1.POP2.fst.idx  -win 20000 -step 20000 -fold 1 > POP1.POP2_fst20k  `

Is anyone else still encountering this error? 

Tony

**Comment by nspope** [comment]
Tony, can you please try using the nspope_bandedDP branch and see if the problem persists? e.g.

`git clone --branch nspope_bandedDP https://github.com/ANGSD/angsd.git && cd angsd && make HTSSRC=/path/to/htslib`

Use `git status` to ensure that you are on the right branch. This branch reworked the way folding is done in realSFS. 

If the numbers still seem odd after this, I think it is worth considering that the implementation is "correct" (insofar as it is calculating the intended estimator), and trying to figure out why different estimators are giving you different answers. What angsd is doing is very simple: for a given site, it calculates the numerator and denominator across all possible allele frequencies, weighting each by the posterior probability that the site has a given frequency. It then sums numerator and denominator across the window, and takes the ratio. This is a super straightforward empirical Bayes estimator. If it gives "wrong" answers, the reason could be: the SFS estimate used as a prior is wonky; the SAF themselves are wonky (this is unlikely unless you used a deprecated option when making the SAF); the windows are too small so the summed numerator/denominator are very noisy.

**Comment by TonyKess** [comment]
I've tried again using the linked branch and continue to get the same results, with both the global weighted and window-based FST estimate too low to be biologically reasonable given results from other marker types, and from different treatment of the same sequencing data (Poolseq, pcangsd). 

I'm now curious what kind of issues might affect the SFS-based FST from my data if the implementation is correct - I have relatively low sequencing depth (1-2x), and few individuals per population in the current comparison. Is it possible that uncertainty from this limited sample is creating high uncertainty in allele frequencies? This pattern does seem similar to what was observed in #274 , which was associated with some sample size differences?

**Comment by TonyKess** [comment]
[2Morph_SFS.pdf](https://github.com/ANGSD/angsd/files/5276509/2Morph_SFS.pdf)

**Comment by TeresaPegan** [comment]
I have been encountering negative Fst values lately also and I am wondering if there are any updates in this situation? I don't have the exact same situation as the original poster, but my sample sizes are very small, so we may be having similar issues?

In fact, I am attempting to calculated Fst between _individuals_ of one large population (for isolation by distance). I am using millions of SNPs. With large numbers of loci, it is possible to calculate Fst between two individuals, e.g. to [calculate an isolation-by-distance slope](https://www.journals.uchicago.edu/doi/pdfplus/10.1086/699515)...at least, you apparently can with an Fst estimator developed by [Reich](https://www.nature.com/articles/nature08365) (see the supplement). However, I am wondering if the Bhatia estimator used by ANGSD with whichfst 1 is not as suitable for this task? Or whether something else might be going on...such as problems with the SFS, like the original poster was wondering about?

 I am doing this with 2d site frequency spectra for each pair of individuals generated by [ngsRelate](https://github.com/ANGSD/NgsRelate) which is a lot easier than ANGSD for making SFS's pairs of individuals (See #413). ngsRelate does not say whether the 2dSFS it makes is folded, but it has 9 elements, just like folded 2dSFS's I've made in ANGSD with pairs of individuals, which makes me think they are calculated in the same way. I made SAFs in ANGSD for each individual and calculated Fst with the Bhatia estimator in ANGSD.

I have tried this on populations from several species. For some species, individual-level pairwise Fst gives me reasonable, positive values and shows a pattern of isolation-by-distance. For other species, the same analysis produces pairwise Fst estimates that are negative for every pair of individuals in the dataset.

Thanks!
-Teresa

**Comment by nspope** [comment]
Hi Teresa, the 1d sfs for a single diploid has three bins (sample freqs of 0, 1, 2) so the joint sfs for two diploids has nine bins. So the ngsRelate output sounds like its not folded. I'd check your calculations.

Otherwise, not sure what to say -- I think the negative values are a result of the bias correction in the numerator of the Bhatia fst estimator (Eq 10 in [this paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3759727/)): the numerator is (p - q)^2 - p*(1-p)/(n - 1) - q*(1-q)/(m-1). p and q are allele frequencies, n and m are sample sizes in the two populations. The last two terms here are the bias correction.

When angsd calculates this estimator, it just takes a weighted average of the numerator and denominator over possible sample allele frequencies, weighting each value by the product of the SFS prior and the SAF likelihood (which is strictly nonnegative). So it will only return a negative value when the estimator itself is negative for some/all allele frequencies. So I'm not sure what to say here aside from a pretty unhelpful "it's because of the estimator".

I guess we could implement an additional estimator that can't go negative -- I think a recent paper by Ragsdale & Gravel (2019 PloS Genetics) does this for a bunch of two-locus statistics, by interpreting the elements of the SFS as raw moments of allele frequencies. It may be awhile before I can get to this though. And I doubt it would change your conclusions (aside from giving non-negative Fst estimates).

**Comment by TeresaPegan** [comment]
Thanks as always for your insights Nate. 
In trying to understand what conditions cause that numerator to be negative (and therefore, what cases the Fst to be negative), I've realized there is something else fundamental that I do not understand: what do the numbers in the SAF file actually mean?
Here are a few lines from a SAF that I have from a single individual (output of realSFS print on the idx file).

``` 
VWZN01000004.1	14235	-4.435523	-0.685288	0.000000
VWZN01000004.1	14241	0.000000	-0.693156	-5.526204
VWZN01000004.1	14245	0.000000	-0.693156	-5.526204
```
I understand that this is the chromosome name and position, followed by 3 numbers which represent the 3 possible allele frequencies in a single individual: 0, 1, or 2. The numbers representing the allele frequencies are all either 0 or negative. Are these likelihoods or something? Does this mean we have AF = 2 in the first locus shown above and AF = 0 in the latter two?
Sorry to have to ask this basic question. I don't think it's in the ANGSD documentation anywhere, but I could be wrong.

**Comment by TeresaPegan** [comment]
Thanks as always for your insights! (Sorry if you got a notification about my previous post, which I deleted, asking about the SAF format -- I found the answer in the [formats.pdf](https://github.com/ANGSD/angsd/blob/master/doc/formats.pdf) in the "docs" folder in this repository).
I have another question as I try to understand what conditions lead to these negative Fsts, but I thought it should probably be its own post (#421) since it is a specific question about how ANGSD works.

**Comment by TeresaPegan** [comment]
Does ANGSD actually use the R scripts in the R folder (fstFrom2dSFS.R and fstFrom2dRHELLER.R) to calculate Fst? If so, it may be fairly doable for me to make my own branch and edit an R file to have it use the Fst estimator I am interested in. But I can't figure out if/how those R scripts are actually used by the program.
Thanks,
-Teresa

**Comment by nspope** [comment]
No, it's done in C++, I think split across a couple of the realSFS_*.cpp
files in angsd/misc. What is the Fst estimator you are interested in?

**Comment by TeresaPegan** [comment]
I would be interested in this one from Reich et al 2009:
https://static-content.springer.com/esm/art%3A10.1038%2Fnature08365/MediaObjects/41586_2009_BFnature08365_MOESM281_ESM.pdf
It's been implemented in python for called SNPs here: https://github.com/singhal/SqCL/blob/master/calculate_divergence_and_Fst.py
(the fst_reich function)
Thanks!
-Teresa
